### DISCUSSION TITLE
fallback page for static exported sites

### DISCUSSION THREAD

**Discussion by LoaiDev** [vercel/next.js#35862] [Ideas]
### Describe the feature you'd like to request

Next js is really powerful even when it is static exported and not just a server. However a big problem with dynamic routes is they wont work when exporting unless you have created `getStaticPaths` however sometimes this isn't possible unless you keep building the website every few minutes. A better way would be to make next manage 404s in a different way. Since all or most webservers can be setup to serve a specific file when not found. this exported file would behave like how a react app would behave.
_Just like `<div id="app"></div>`_

### Describe the solution you'd like

have all of the 404 requests be handled first through the next router to check for dynamic routes and then fallback to the 404 page if it is not a dynamic route. This will make static exports much more powerful with still having all the other static pages pre rendered.

### Describe alternatives you've considered

we could manually edit the 404 page to do all of this but it would be more complicated.

**Comment by carlnc** [comment]
Hello @LoaiDev 

Looking at the output of `next export`'s `build/` folder:


1. The produced `index.html` pages contain a unique footer that dictates the page to render (see below).
2. The file `_buildManifest.js` (under `build/_next/static/…/`) contains a list of all the routes, and the JS files that need to be loaded to handle/render that page.

```html
    <!-- from index.html / [resourceId].html file -->

    <script id="__NEXT_DATA__" type="application/json">{
      "props": {
        "pageProps": {
          "statusCode": 404
        }
      },
      "page": "/path/to/some/resource/[resourceId]",  <<<<<<<< THIS HERE
      "query": { },
      "buildId": "KWMalEB3hB-Ya5uRudhT0",
      "nextExport": true,
      "isFallback": false,
      "gip": true,
      "scriptLoader": []
    }</script>
```

```javascript
    // from _buildManifest.js
    return {
        "/":                                                ["static/chunks/pages/index-b5903b905e5cf194.js"],
        "/_error":                                          ["static/chunks/pages/_error-9ba256efc4caf4b0.js"],
        "/path/to/some/resource/[resourceId]":              [
            I,
            T,
            "static/chunks/pages/path/to/some/resource/[resourceId]-9782bd746bb74d83.js",
        ],
```

An alternative, for NextJS, would have been to produce a single `index.html` (duplicated everywhere), that instead uses `window.location` (not `__NEXT_DATA__.page`) to figure out which `_buildManifest.js` item to use.

so: 
* window.location  = `https://your-site.com/path/to/some/resource/123`
* -> resource path = `/path/to/some/resource/123`
* -> match in _buildManifest = `/path/to/some/resource/[resourceId]`

and if no match is found: render _buildManifest item `/_error` (ie: the 404 page)

**Comment by LoaiDev** [comment]
I think you didn't understand my suggestion correctly. say a user visits example.com/post/32. now he will be shown a 404 page because 32 doesn't exist in getStaticPaths. while it would work when using a node server it now won't work on ssg. So my idea is before showing the 404 page, the client should first check if a path exists in the router first so that when it checks it finds the the /post/[id] and renders the post page instead of a 404. This will all be rendered on the client side which is a downside but it will make any dynamic route work when exporting.

**Comment by carlnc** [comment]
@LoaiDev 

Please clarify what you mean by `before showing the 404 page`

* before the server renders `{build}/404/index.html`,
* before `{build}/404/index.html` renders the `{src}/pages/404.js`
?

**Reply by LoaiDev** [reply]
The second case. where the client itself checks before rendering the 404.js page

**Reply by LoaiDev** [reply]
so the exported 404.html would have some sort of checking the router before rendering the 404.js src page. this way dynamic routes would still work while also having it being served by a cdn without a server

**Reply by carlnc** [reply]
@LoaiDev 

OK.  I'm not sure how my suggestion isn't a superset of your suggestion.

If I understand correctly, what you have suggested for `{build}/404/index.html`, is what I have suggested for very `index.html`.

**Reply by LoaiDev** [reply]
yes I think your suggestion is pretty much the same as what I suggested. I just got a little confused because I thought you were explaining something rather than giving your understanding of what I mean

**Comment by carlnc** [comment]
@LoaiDev 

As a work around we use query params, and do the routing ourselves (ie: `/post/?id=123` (which is gross))

So instead of `/post/[id].tsx` we have `/post/index.tsx`.

and in `/post/index.tsx`:

```typescript
export default function PostsRouter(): JSX.Element {
    const router = useRouter();
    const postId = router.query.id;

    return postId
        ? <PostsPage />
        : <PostsList />;
}
```

** Note: we have `trailingSlash: true` turned on, so your filenames might change.

**Comment by LoaiDev** [comment]
This workaround works but it's not as clean as using this feature. Do you think it is a good feature and fit for next js? It is kinda a dirty solution as well to this problem but it is better than the alternative and would also require no changes if the developer decides to move from static pages to a server or vice versa. What are the chances of this ever going into the main framework?

**Comment by carlnc** [comment]
@LoaiDev 

> This workaround works but it's not as clean as using this feature.

Yes, it goes against standard practice and it looks unprofessional.

> Do you think it is a good feature and fit for next js? 

Upon testing, I now notice that transitioning from one page to another using `router.push()` results in a full page load.  I was of the understanding/expectation that, once a page was loaded, page transitions behaved like a SPA.  My solution was to piggyback off of that.

I also don't see supporting "dynamic routes for static pages" in the repository owner's interest.


We're using `Cloudfront & S3`, so we might look into writing a Lamdba to translate dynamic routes (using the contents of _buildManifest.js).

```javascript
global.self = {};
require('./build/_next/static/n-O0UUVdWhmmEJG1VqqzP/_buildManifest');
console.log(self.__BUILD_MANIFEST.sortedPages);
```
